gson/src/main/java/com/google/gson/internal/bind/ReflectiveTypeAdapterFactory.java

```diff
@@ -215,11 +221,24 @@ protected BoundField(String name, boolean serialized, boolean deserialized) {
         in.beginObject();
         while (in.hasNext()) {
           String name = in.nextName();
-          BoundField field = boundFields.get(name);
-          if (field == null || !field.deserialized) {
-            in.skipValue();
+          if (useJsog && JsogRegistry.JSOG_FIELD_ID.equals(name)) {
+            // JsogEnabled Id -- Registers the instance on JsogEnabled
+            String jsogId = in.nextString();
+            JsogRegistry.get().register(jsogId, instance);
+          } else if (useJsog && JsogRegistry.JSOG_FIELD_REF.equals(name)) {
+            // JsogEnabled backreference -- Replaces the instance with the value registered on JsogEnabled
+            String jsogId = in.nextString();
+            instance = (T) JsogRegistry.get().getInstance(jsogId);
+            if (instance == null) {
+              throw new JsonSyntaxException("Unresolved JSOG reference: " + jsogId);
+            }
```

Comment:
This should verify that `hasNext()` returns `false` and if not throw an exception. Otherwise it would modify an existing object from the registry. (Unless that is a JSOG feature, which I doubt).

Maybe this should be solved differently in general since it would unnecessarily call `constructor.construct();` just to discard it then and would not fail if there are other non `@ref` properties before, modifying the existing `instance` just to discard it then.
Maybe at the beginning read one name (if there is one) check if it is `@ref`, otherwise store in local variable `nextName` and use that in loop and at the end of the loop assign `nextName = in.nextName()` (if there is one).